sdk/typescript/src/utils/gas-estimation.ts

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { UnserializedSignableTransaction } from '../signers/txn-data-serializers/txn-data-serializer';
+import { Coin, SuiMoveObject } from '../types';
+
+export type TxKind = UnserializedSignableTransaction['kind'];
+type BudgetMap = typeof DEFAULT_GAS_BUDGET_PER_TX_TYPE;
+export type GasBudgetGuessParams<txKind extends TxKind> =
+  txKind extends keyof BudgetMap
+    ? BudgetMap[txKind] extends (...args: any[]) => any
+      ? Parameters<BudgetMap[txKind]> extends [boolean, ...infer Params]
+        ? Params
+        : []
+      : []
+    : [];
+
+export const MIN_GAS_BUDGET = 10;
+export const MAX_GAS_BUDGET = 1_000_000;
+
+const PAY_GAS_FEE_PER_COIN = 150;
+
+/**
+ * Make a guess for the gas budget of a pay tx
+ * @param coins All the coins of the type to send
+ * @param amount The amount to send
+ * @returns The gasBudget guess
+ */
+function computeTransferTxGasBudget(
+  optimizeGuessForDryRun: boolean,
+  coins: SuiMoveObject[],
+  amount: bigint
+) {
+  const numInputCoins = Coin.selectCoinSetWithCombinedBalanceGreaterThanOrEqual(
+    coins,
+    amount
+  ).length;
+  let gasBudgetGuess = Math.min(
+    PAY_GAS_FEE_PER_COIN * Math.max(2, Math.min(100, numInputCoins / 2)),
+    MAX_GAS_BUDGET
+  );
+  const isSuiTransfer = Coin.isSUI(coins[0]);
+  if (isSuiTransfer && optimizeGuessForDryRun) {
+    // check if there is enough balance to cover for the amount + the gas
+    // if not lower the gasBudget to allow making better estimations for those cases
+    const totalSuiBalance = coins.reduce(
+      (sum, aCoin) => sum + (Coin.getBalance(aCoin) || BigInt(0)),
+      BigInt(0)
+    );
+    if (totalSuiBalance - BigInt(gasBudgetGuess) - amount < 0) {
+      gasBudgetGuess = Math.max(
+        Number(totalSuiBalance - amount),
+        MIN_GAS_BUDGET
+      );
+    }
+  }
+  return gasBudgetGuess;
+}
+
+const DEFAULT_GAS_BUDGET_PER_TX_TYPE = {
+  mergeCoin: 10_000,
+  moveCall: 10_000,
+  pay: computeTransferTxGasBudget,
+  paySui: computeTransferTxGasBudget,
+  payAllSui: computeTransferTxGasBudget,
+  publish: 10_000,
+  splitCoin: 10_000,
+  transferObject: 100,
+  transferSui: 100,
+} as const;
+
+export function getGasBudgetGuess<T extends TxKind>(
```

Comment:
Why can't this use dry-run?

Reply:
Why not just set it to something very high when dry-running?

@gegaowp I'm curious if we can still succeed for dry run even when gas budget is exceeded.

Reply:
> Why not just set it to something very high when dry-running

Then we have the problem that tx can fail because of insufficient funds, but still we need to handle the case dry-run fails. We could simply set it to the value of the coin with max balance (-amount in case we send Sui, etc) which is what we are doing except we start from a max gas budget that makes sense for the tx type

Reply:
This makes more sense to me. I really do think we need to fix this at the dry run level, and not paper over it like this. I definitely don’t think these should be in the SDK, as ideally we can just fix dry run. As it stands I think it’s way too complicated for partners to actually estimate gas, and we should fix that. I read the comments on the other thread too, seems like something that is solvable. I think we should even if it’s hard. 

I get the case where the dry run fails. I think in those cases we either use a default max value, and warn the user their transaction may fail, and that it may use a lot of gas. IMO we need to guess it for those cases, we just need to tell the user. Ideally dry run failing is actually our signal that the tx itself will fail and we’ll actually gate the default transaction UI on that.

Reply:
I agree changing dry-run to ignore gas budget would help, the only thing that will change though would be the removal of the `optimizeForDryRun` and probably the `maxGasCoinBalance` of this function. We still need for some cases the guess constant to be a function, because gas depends on number of coins etc.

This function is more meant as an internal to the sdk utility - users should use the [getGasCostEstimationAndSuggestedBudget](https://github.com/MystenLabs/sui/pull/6156/files/5bd6e33ba5f0556ca90cb5ef3bf1feffbf56be30#diff-90a054a3b632795af85582441a61df5784805807e652e8167d61978a198a17faR385) of the signer where it returns the estimation (either from dry-run or from guess).

We can easily remove the extra functionality, to optimise for dry-run when dry run changes and that should not affect much the end user.

I think [getGasCostEstimationAndSuggestedBudget](https://github.com/MystenLabs/sui/pull/6156/files/5bd6e33ba5f0556ca90cb5ef3bf1feffbf56be30#diff-90a054a3b632795af85582441a61df5784805807e652e8167d61978a198a17faR385) works more or less how you describe above, maybe it need some changes to it's api.